// resulty/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromEmpy = exports.fromNullable = exports.Result = exports.ok = exports.Ok = exports.err = exports.Err = void 0;
var Err_1 = __importStar(require("./Err"));
exports.Err = Err_1.default;
Object.defineProperty(exports, "err", { enumerable: true, get: function () { return Err_1.err; } });
var Ok_1 = __importStar(require("./Ok"));
exports.Ok = Ok_1.default;
Object.defineProperty(exports, "ok", { enumerable: true, get: function () { return Ok_1.ok; } });
var Result_1 = __importDefault(require("./Result"));
exports.Result = Result_1.default;
var fromNullable = function (errorCase, x) {
    return x ? Ok_1.ok(x) : Err_1.err(errorCase);
};
exports.fromNullable = fromNullable;
var fromEmpy = function (errorCase, xs) {
    return xs.length === 0 ? Err_1.err(errorCase) : Ok_1.ok(xs);
};
exports.fromEmpy = fromEmpy;
