// maybeasy/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNothing = exports.isJust = exports.fromEmpty = exports.fromNullable = exports.nothing = exports.Nothing = exports.just = exports.Just = exports.Maybe = void 0;
var Just_1 = require("./Just");
exports.Just = Just_1.default;
Object.defineProperty(exports, "just", { enumerable: true, get: function () { return Just_1.just; } });
var Maybe_1 = require("./Maybe");
exports.Maybe = Maybe_1.default;
var Nothing_1 = require("./Nothing");
exports.Nothing = Nothing_1.default;
Object.defineProperty(exports, "nothing", { enumerable: true, get: function () { return Nothing_1.nothing; } });
var fromNullable = function (v) {
    if (v) {
        return Just_1.just(v);
    }
    return Nothing_1.nothing();
};
exports.fromNullable = fromNullable;
var fromEmpty = function (xs) {
    return xs.length === 0 ? Nothing_1.nothing() : Just_1.just(xs);
};
exports.fromEmpty = fromEmpty;
var isJust = function (maybe) { return maybe instanceof Just_1.default; };
exports.isJust = isJust;
var isNothing = function (maybe) { return maybe instanceof Nothing_1.default; };
exports.isNothing = isNothing;
