// maybeasy/lib/Nothing.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.nothing = exports.Nothing = void 0;
var Maybe_1 = require("./Maybe");
var Nothing = (function (_super) {
    __extends(Nothing, _super);
    function Nothing() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Nothing.prototype.getOrElse = function (fn) {
        return fn();
    };
    Nothing.prototype.getOrElseValue = function (defaultValue) {
        return defaultValue;
    };
    Nothing.prototype.map = function (fn) {
        return new Nothing();
    };
    Nothing.prototype.andThen = function (fn) {
        return new Nothing();
    };
    Nothing.prototype.orElse = function (fn) {
        return fn();
    };
    Nothing.prototype.cata = function (matcher) {
        return matcher.Nothing();
    };
    Nothing.prototype.assign = function (k, other) {
        return new Nothing();
    };
    Nothing.prototype.do = function (fn) {
        return new Nothing();
    };
    Nothing.prototype.elseDo = function (fn) {
        fn();
        return this;
    };
    return Nothing;
}(Maybe_1.default));
exports.Nothing = Nothing;
var nothing = function () { return new Nothing(); };
exports.nothing = nothing;
exports.default = Nothing;
