// resulty/lib/Err.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.err = void 0;
var Result_1 = __importDefault(require("./Result"));
var Err = (function (_super) {
    __extends(Err, _super);
    function Err(error) {
        var _this = _super.call(this) || this;
        _this.error = error;
        return _this;
    }
    Err.prototype.getOrElse = function (fn) {
        return fn();
    };
    Err.prototype.getOrElseValue = function (defaultValue) {
        return defaultValue;
    };
    Err.prototype.map = function (fn) {
        return new Err(this.error);
    };
    Err.prototype.mapError = function (fn) {
        return new Err(fn(this.error));
    };
    Err.prototype.andThen = function (fn) {
        return new Err(this.error);
    };
    Err.prototype.orElse = function (fn) {
        return fn(this.error);
    };
    Err.prototype.cata = function (matcher) {
        return matcher.Err(this.error);
    };
    Err.prototype.ap = function (result) {
        return new Err(this.error);
    };
    Err.prototype.assign = function (k, other) {
        return new Err(this.error);
    };
    Err.prototype.do = function (fn) {
        return new Err(this.error);
    };
    Err.prototype.elseDo = function (fn) {
        fn(this.error);
        return this;
    };
    return Err;
}(Result_1.default));
var err = function (e) { return new Err(e); };
exports.err = err;
exports.default = Err;
